cloud/gke.go

```diff
@@ -0,0 +1,1080 @@
+	if cc.NodePoolID == "" {
+		cluster, err := svc.Projects.Zones.Clusters.Get(cc.ProjectID, cc.Zone, cc.Name).Context(context.Background()).Do()
+		if err != nil {
+			banzaiUtils.LogError(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+		if cluster.NodePools != nil && len(cluster.NodePools) != 0 {
+			cc.NodePoolID = cluster.NodePools[0].Name
+		}
+	}
+
+	if cc.MasterVersion != "" {
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Updating master to %v version", cc.MasterVersion)
+		updateCall, err := svc.Projects.Zones.Clusters.Update(cc.ProjectID, cc.Zone, cc.Name, &gke.UpdateClusterRequest{
+			Update: &gke.ClusterUpdate{
+				DesiredMasterVersion: cc.MasterVersion,
+			},
+		}).Context(context.Background()).Do()
+		if err != nil {
+			return nil, err
+		}
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Cluster %s update is called for project %s and zone %s. Status Code %v", cc.Name, cc.ProjectID, cc.Zone, updateCall.HTTPStatusCode)
+		if updatedCluster, err = waitForCluster(svc, cc); err != nil {
+			banzaiUtils.LogError(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+	}
+
+	if cc.NodeVersion != "" {
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Updating node to %v verison", cc.NodeVersion)
+		updateCall, err := svc.Projects.Zones.Clusters.NodePools.Update(cc.ProjectID, cc.Zone, cc.Name, cc.NodePoolID, &gke.UpdateNodePoolRequest{
+			NodeVersion: cc.NodeVersion,
+		}).Context(context.Background()).Do()
+		if err != nil {
+			banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Nodepool %s update is called for project %s, zone %s and cluster %s. Status Code %v", cc.NodePoolID, cc.ProjectID, cc.Zone, cc.Name, updateCall.HTTPStatusCode)
+		if err := waitForNodePool(svc, cc); err != nil {
+			banzaiUtils.LogError(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+	}
+
+	if cc.NodeCount != 0 {
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Updating node size to %v", cc.NodeCount)
+		updateCall, err := svc.Projects.Zones.Clusters.NodePools.SetSize(cc.ProjectID, cc.Zone, cc.Name, cc.NodePoolID, &gke.SetNodePoolSizeRequest{
+			NodeCount: cc.NodeCount,
+		}).Context(context.Background()).Do()
+		if err != nil {
+			return nil, err
+		}
+		banzaiUtils.LogInfof(banzaiConstants.TagUpdateCluster, "Nodepool %s size change is called for project %s, zone %s and cluster %s. Status Code %v", cc.NodePoolID, cc.ProjectID, cc.Zone, cc.Name, updateCall.HTTPStatusCode)
+		if updatedCluster, err = waitForCluster(svc, cc); err != nil {
+			banzaiUtils.LogError(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+	}
+	return updatedCluster, nil
+}
+
+func waitForNodePool(svc *gke.Service, cc GKECluster) error {
+	const TAG = "waitForNodePool"
+	message := ""
```

Comment:
Just a minor thing I think we should use var message string here instead of adding empty string.